homeassistant/components/risco/alarm_control_panel.py

```diff
@@ -117,32 +132,49 @@ def code_format(self):
         """Return one or more digits/characters."""
         return FORMAT_NUMBER
 
-    def _validate_code(self, code, state):
+    def _validate_code(self, code):
         """Validate given code."""
-        check = code == self._code
-        if not check:
-            _LOGGER.warning("Wrong code entered for %s", state)
-        return check
+        return code == self._code
 
     async def async_alarm_disarm(self, code=None):
         """Send disarm command."""
-        if self._code_disarm_required and not self._validate_code(code, "disarming"):
+        if self._code_disarm_required and not self._validate_code(code):
+            _LOGGER.warning("Wrong code entered for disarming")
             return
         await self._call_alarm_method("disarm")
 
     async def async_alarm_arm_home(self, code=None):
         """Send arm home command."""
-        if self._code_arm_required and not self._validate_code(code, "arming home"):
-            return
-        await self._call_alarm_method("partial_arm")
+        await self._arm(STATE_ALARM_ARMED_HOME, code)
 
     async def async_alarm_arm_away(self, code=None):
         """Send arm away command."""
-        if self._code_arm_required and not self._validate_code(code, "arming away"):
+        await self._arm(STATE_ALARM_ARMED_AWAY, code)
+
+    async def async_alarm_arm_night(self, code=None):
+        """Send arm night command."""
+        await self._arm(STATE_ALARM_ARMED_NIGHT, code)
+
+    async def async_alarm_arm_custom_bypass(self, code=None):
+        """Send arm custom bypass command."""
+        await self._arm(STATE_ALARM_ARMED_CUSTOM_BYPASS, code)
+
+    async def _arm(self, mode, code):
+        if self._code_arm_required and not self._validate_code(code):
+            _LOGGER.warning("Wrong code entered for %s", mode)
+            return
+
+        risco_state = self._ha_to_risco[mode]
+        if not risco_state:
+            _LOGGER.warning("No mapping for mode %s", mode)
             return
```

Comment:
Not sure how to cover this with a test. Hypothetically, this condition can't happen, but it seems wise to add this guard to protect against a bug that would cause it to.

Reply:
have you tried initializing the mock entity with options that are missing one of the keys? might that trigger it?

Reply:
Any keys that are missing in `ha_to_risco` mean that the entity wouldn't report those services as supported, thus there's no way to run them...